SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestReplicationConditions/SpatialTestReplicationConditions.cpp

```diff
@@ -0,0 +1,604 @@
+		TestActor_Common->SetOwner(PlayerController);
+
+		TestActor_AutonomousOnly->SetOwner(PlayerController);
+		TestActor_AutonomousOnly->SetAutonomousProxy(true);
+
+		// Set both physics actor to autonomous so properties won't be replicated through simulated condition
+		TestActor_PhysicsEnabled->SetOwner(PlayerController);
+		TestActor_PhysicsDisabled->SetOwner(PlayerController);
+		TestActor_PhysicsEnabled->SetAutonomousProxy(true);
+		TestActor_PhysicsDisabled->SetAutonomousProxy(true);
+
+		RegisterAutoDestroyActor(TestActor_Common);
+		RegisterAutoDestroyActor(TestActor_CustomEnabled);
+		RegisterAutoDestroyActor(TestActor_CustomDisabled);
+		RegisterAutoDestroyActor(TestActor_AutonomousOnly);
+		RegisterAutoDestroyActor(TestActor_PhysicsEnabled);
+		RegisterAutoDestroyActor(TestActor_PhysicsDisabled);
+
+		FinishStep();
+	});
+
+	// Check on non-auth server that actor exists and properties are correct
+	if (bSpatialEnabled)
+	{
+		AddStepFromDefinition(CheckOnNonAuthServerStepDefinition, FWorkerDefinition::Server(2));
+	}
+
+	// Check on owning client that actor exists and properties are correct
+	AddStepFromDefinition(CheckOnOwningClientStepDefinition, FWorkerDefinition::Client(1));
+
+	// Check on non-owning client that actor exists and properties are correct
+	AddStepFromDefinition(CheckOnNonOwningClientStepDefinition, FWorkerDefinition::Client(2));
+
+	AddStep(TEXT("ASpatialTestReplicationConditions Update Test Actors To Expect New Properties"), FWorkerDefinition::AllWorkers, nullptr,
+			[this]() {
+				ReplicationStage = STAGE_UpdateReplication;
+				PropertyOffset = 10000;
+
+				FinishStep();
+			});
+
+	// The Server updates TestActors with new replicated properties
+	AddStep(TEXT("ASpatialTestReplicationConditions Update Test Actors With New Properties"), FWorkerDefinition::Server(1), nullptr,
+			[this]() {
+				const bool bWrite = true;
+				bool bCondIgnore[COND_Max]{};
+				ProcessCommonActorProperties(bWrite, bCondIgnore);
+
+				ProcessCustomActorProperties(TestActor_CustomEnabled, bWrite, /*bCustomEnabled*/ true);
+				ProcessCustomActorProperties(TestActor_CustomDisabled, bWrite, /*bCustomEnabled*/ false);
+
+				ProcessAutonomousOnlyActorProperties(bWrite, /*bAutonomousExpected*/ false, /*bSimulatedExpected*/ false);
+
+				ProcessPhysicsActorProperties(TestActor_PhysicsEnabled, bWrite, /*bPhysicsEnabled*/ true, /*bPhysicsExpected*/ true);
+				ProcessPhysicsActorProperties(TestActor_PhysicsDisabled, bWrite, /*bPhysicsEnabled*/ false, /*bPhysicsExpected*/ false);
+
+				FinishStep();
+			});
+
+	// Check on non-auth server that actor exists and properties are correct
+	if (bSpatialEnabled)
+	{
+		AddStepFromDefinition(CheckOnNonAuthServerStepDefinition, FWorkerDefinition::Server(2));
+	}
+
+	// Check on owning client that actor exists and properties are correct
+	AddStepFromDefinition(CheckOnOwningClientStepDefinition, FWorkerDefinition::Client(1));
```

Comment:
By the way, does this not flake? The function is doing asserts, but here we are running on the client right after changing the variables on the server.
Is there a guarantee the variables will arrive to the clients by this time?

Reply:
This doesn't actually, perhaps the progression of the step means that the variable is always replicated?

Reply:
Unfortunately, I think this is a race condition.
Have you tried running this with the replication graph? It seems like ReplicationGraph processes RPCs before variable replication, so this could fail there much more often.

Reply:
Unfortunately fix is probably not that easy - would probably have to switch to requires, and reason about whether that's correct, or worst case add a time delay.

Reply:
Will switch to requires.